src/probability_theory/density.lean

```diff
@@ -0,0 +1,380 @@
+
+## Main definitions
+
+* `measure_theory.measure.has_pdf` : A random variable `X : α → E` is said to `has_pdf` with
+  respect to the measure `ℙ` on `α` and `μ` on `E` if there exists a measurable function `f`
+  such that the push-forward measure of `ℙ` along `X` equals `μ.with_density f`.
+* `measure_theory.measure.pdf` : If `X` is a random variable that `has_pdf X ℙ μ`, then `pdf X`
+  is the measurable function `f` such that the push-forward measure of `ℙ` along `X` equals
+  `μ.with_density f`.
+* `measure_theory.measure.pdf.uniform` : A random variable `X` is said to follow the uniform
+  distribution if it has a constant probability density function with a compact, non-null support.
+
+## Main results
+
+* `measure_theory.measure.pdf.integral_mul_eq_integral'` : Law of the unconscious statistician,
+  i.e. if a random variable `X : α → E` has pdf `f`, then `𝔼(g(X)) = ∫ x, g x * f x dx` for
+  all measurable `g : E → ℝ`.
+* `measure_theory.measure.pdf.integral_mul_eq_integral` : A real-valued random variable `X` with
+  pdf `f` has expectation `∫ x, x * f x dx`.
+* `measure_theory.measure.pdf.uniform.integral_eq` : If `X` follows the uniform distribution with
+  its pdf having support `s`, then `X` has expectation `(λ s)⁻¹ * ∫ x in s, x dx` where `λ`
+  is the Lebesgue measure.
+
+## TODOs
+
+Ultimately, we would also like to define characteristic functions to describe distributions as
+it exists for all random variables. However, to define this, we will need Fourier transforms
+which we currently do not have.
+-/
+
+noncomputable theory
+open_locale classical measure_theory nnreal ennreal
+
+namespace measure_theory
+
+open topological_space
+
+variables {α E : Type*} [normed_group E] [measurable_space E] [second_countable_topology E]
+  [normed_space ℝ E] [complete_space E] [borel_space E]
+
+namespace measure
+
+/-- A random variable `X : α → E` is said to `has_pdf` with respect to the measure `ℙ` on `α` and
+`μ` on `E` if there exists a measurable function `f` such that the push-forward measure of `ℙ`
+along `X` equals `μ.with_density f`. -/
+class has_pdf {m : measurable_space α} (X : α → E)
+  (ℙ : measure α) (μ : measure E . volume_tac) : Prop :=
+(pdf' : ∃ (f : E → ℝ≥0∞), measurable f ∧ map X ℙ = μ.with_density f)
+
+/-- If `X` is a random variable that `has_pdf X ℙ μ`, then `pdf X` is the measurable function `f`
+such that the push-forward measure of `ℙ` along `X` equals `μ.with_density f`. -/
+def pdf {m : measurable_space α} (X : α → E) (ℙ : measure α) (μ : measure E . volume_tac) :=
+if hX : has_pdf X ℙ μ then classical.some hX.pdf' else 0
+
+@[measurability]
+lemma measurable_pdf {m : measurable_space α}
+  (X : α → E) (ℙ : measure α) (μ : measure E . volume_tac) :
+  measurable (pdf X ℙ μ) :=
+begin
+  by_cases hX : has_pdf X ℙ μ,
+  { rw [pdf, dif_pos hX],
+    exact (classical.some_spec hX.pdf').1 },
+  { rw [pdf, dif_neg hX],
+    exact measurable_zero }
+end
+
+lemma pdf_spec {m : measurable_space α}
```

Comment:
A more descriptive name would be nicer. `map_eq_with_density_pdf`?

Same remark for `pdf_spec'`.